quil-py/quil/program/__init__.pyi

```diff
@@ -238,6 +314,73 @@ class CalibrationSet:
         """
         ...
 
+
+@final
+class FixedScheduleItem:
+    """
+    A single item within a fixed schedule, representing a single instruction within a basic block.
+    """
+    @property
+    def instruction_index(self) -> int:
+        """
+        The index of the instruction within the basic block.
+        """
+    @property
+    def time_span(self) -> FixedTimeSpan:
+        """
+        The time span during which the instruction is scheduled.
+        """
+
+
+@final
+class ControlFlowGraph:
+    """
+    Representation of a control flow graph (CFG) for a Quil program.
+
+    The CFG is a directed graph where each node is a basic block and each edge is a control flow
+    transition between two basic blocks.
+    """
+
+    def has_dynamic_control_flow(self) -> bool: 
+        """
+        Return ``True`` if the program has dynamic control flow, i.e. contains a conditional branch instruction.
+        """
+    def basic_blocks(self) -> List["BasicBlock"]: 
+        """
+        Return a list of all the basic blocks in the control flow graph, in order of definition.
+        """
+
+@final
+class FixedSchedule:
+    def items(self) -> List[FixedScheduleItem]:
+        """
+        All the items in the schedule, in unspecified order.
+        """
+    def duration(self) -> float:
+        """
+        The duration of the schedule, in seconds.
+
+        This is the maximum of the end time of all the items.
+        """
+
+@final
+class FixedTimeSpan:
+    """
+    Representation of a time span in seconds.
+    """
+    @property
+    def start(self) -> float:
```

Comment:
I completely understand why you would use `float`, in keeping `quil-rs` hardware agnostic. However, given the typical durations will be nanosecond or microsecond, and printing floats in a typical use case results in e.g. "5.6000000000000005e-08 to 9.600000000000001e-08" being the output, where do you see the practical issue of latching to some nanosecond or microsecond hardware "tick" solved?

Reply:
This PR doesn't really make a decision there; durations are today already specified in floating-point seconds within the waveform templates, so this new feature just accumulates them. That is, if you were to round-trip `PULSE 0 "rf" flat(duration: 5.6e-8)` on `master`, you'd encounter the same floating-point rounding error.

They are latched by the consumer of the schedule, who can also generate a `Schedule<TimeUnit>` using a `TimeUnit` of their choice.

Reply:
OK cool. Doing something like latching to a ns and getting an integer could be interesting. Future feature?